language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+    desc "The server providing the remote filebucket service."
+    defaultto { Puppet[:server] }
+  end
+end
+```
+
+The default value of `Puppet[:server]` works for all agents, because the default value is resolved at catalog application time, not compilation time.
+
+### is_to_s/should_to_s/change_to_s
+
+Overrides log messages for a property at catalog application time. Common examples are convert an id to a human readable name, sorting multi-valued attributes, redacting passwords, etc.
+
+```ruby
+Puppet::Type.newtype(:user) do
+  newproperty(:password) do
+    def is_to_s(currentvalue)
+      return '[old password hash redacted]'
+    end
+  end
+end
+```
+
+## Parameter/Property Options
+
+The following options are passed to the `newparam` and `newproperty` methods to modify their behavior. These are largely hacks because puppet's predefined parameter/properties types are not well-specified or complete. For example, there isn't a `Puppet::Parameter::Integer` class.
+
+### array_matching
+
+By default, if multiple desired (aka `should`) values are specified in a manifest, puppet will make sure the current (aka `is`) value matches at least one of the desired values. Alternatively, you can specify `array_matching => all`, and puppet will ensure that the array of current values match the desired values.
+
+By default, the `insync?` comparison is sensitive to order and duplicate values, so sometimes a type will override the `insync?` method. For example, security groups are compared set-wise, which ignores duplicates.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:security_groups, :array_matching => :all) do
+    desc 'The security groups to associate the instance.'
+    def insync?(is)
+      is.to_set == should.to_set
+    end
+  end
+end
+```
+
+### parent
+
+Specifies a parent class that the newly defined property/parameter should extend. For example, the `ec2_instance` type defines a `tags` property that inherits from the `PuppetX::Property::AwsTag` class:
+
+```ruby
+require_relative '../../puppet_x/puppetlabs/property/tag.rb'
+
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:tags, :parent => PuppetX::Property::AwsTag) do
+    desc 'The tags for the instance.'
+  end
+end
+```
+
+The `PuppetX::Property::AwsTag` class is defined in helper code, so the type must require it using a relative path. The helper code defines logic for validating aws tags and how log messages are written:
+
+```ruby
+module PuppetX
+  module Property
+    class AwsTag < Puppet::Property
+
+      def format_tags(value)
+        Hash[value.sort]
+      end
+
+      [:should_to_s, :is_to_s].each { |method|
+        alias_method method, :format_tags
+      }
+
+      validate do |value|
+        fail 'tags should be a Hash' unless value.is_a?(Hash)
+      end
+    end
+  end
+end
+```
+
+###  boolean
+
+Meta-programs a predicate method for the parameter/property. For example, the file type defines a `force` parameter:
```

Comment:
There should be support for several different property types beyond boolean array to handle whether items are inclusive, does order matter, array of hashes, hashes etc: https://github.com/vmware/vmware-vmware_lib/tree/master/lib/puppet/property